src/thermo/ThermoPhase.cpp

```diff
@@ -689,6 +689,617 @@ void ThermoPhase::setState_SPorSV(double Starget, double p,
+}
+
+void ThermoPhase::setMixtureFraction_Y(double mixFrac, const compositionMap& fuelComp,
+                                       const compositionMap& oxComp)
+{
+    vector_fp Yf(getCompositionFromMap(fuelComp));
+    vector_fp Yo(getCompositionFromMap(oxComp));
+    return setMixtureFraction_Y(mixFrac, Yf.data(), Yo.data());
+}
+
+void ThermoPhase::setMixtureFraction_Y(double mixFrac, const std::string& fuelComp,
+                                       const std::string& oxComp)
+{
+    setMixtureFraction_Y(mixFrac,
+        parseCompString(fuelComp.find(":") != std::string::npos ? fuelComp : fuelComp+":1.0"),
+        parseCompString(oxComp.find(":") != std::string::npos ? oxComp : oxComp+":1.0"));
+}
+
+void ThermoPhase::setMixtureFraction_Y(double mixFrac, const vector_fp& fuelComp,
+                                       const vector_fp& oxComp)
+{
+    setMixtureFraction_Y(mixFrac, fuelComp.data(), oxComp.data());
+}
+
+void ThermoPhase::setMixtureFraction_Y(double mixFrac, const double* fuelComp,
+                                       const double* oxComp)
+{
+    if (mixFrac < 0.0 || mixFrac > 1.0)
+        throw CanteraError("ThermoPhase::setMixtureFraction_Y",
+                           "mixture fraction must be between 0 and 1");
+
+    auto sum_Yf = std::accumulate(fuelComp, fuelComp+m_kk, 0.0);
+    auto sum_Yo = std::accumulate(oxComp, oxComp+m_kk, 0.0);
+    
+    if (sum_Yf == 0.0 || sum_Yo == 0.0)
+        throw CanteraError("ThermoPhase::setMixtureFraction_Y",
+                           "no fuel and/or oxidizer composition specified");
+
+    auto p = pressure();
+    
+    vector_fp Y(m_kk);
+    
+    for (size_t i=0; i!=m_kk; ++i)
+        Y[i] = mixFrac * fuelComp[i]/sum_Yf + (1.0-mixFrac) * oxComp[i]/sum_Yo;
+    
+    setMassFractions_NoNorm(Y.data());
+    setPressure(p);
+}
+
+double ThermoPhase::getMixtureFraction_X(const compositionMap& fuelComp,
+                                         const compositionMap& oxComp) const
+{
+    vector_fp Xf(getCompositionFromMap(fuelComp));
+    vector_fp Xo(getCompositionFromMap(oxComp));
+    return getMixtureFraction_X(Xf.data(), Xo.data());
+}
+
+double ThermoPhase::getMixtureFraction_X(const std::string& fuelComp,
+                                         const std::string& oxComp) const
+{
+    return getMixtureFraction_X(
+            parseCompString(fuelComp.find(":") != std::string::npos ? fuelComp : fuelComp+":1.0"),
+            parseCompString(oxComp.find(":") != std::string::npos ? oxComp : oxComp+":1.0"));
+}
+
+double ThermoPhase::getMixtureFraction_X(const vector_fp& fuelComp,
+                                         const vector_fp& oxComp) const
+{
+    return getMixtureFraction_X(fuelComp.data(), oxComp.data());
+}
+
+double ThermoPhase::getMixtureFraction_X(const double* fuelComp,
+                                        const double* oxComp) const
+{
+    // mass fractions of fuel and oxidizer
+    vector_fp massFractions(m_kk * 2);
```

Comment:
Using a single vector to store both the fuel and oxidizer compositions is an unnecessary complication.